### ISSUE TITLE
defense against timewarp attack

### ISSUE THREAD

**Issue by zookozcash** [zcash/zcash#696]
https://github.com/bitcoin/bips/blob/master/bip-0099.mediawiki

Seems to say that timewarp attack was used in practice against altcoins. This is news to me because I thought it wasn't easily exploited in practice, so I asked for clarification on #bitcoin-wizards:

```
<Luke-Jr> Thoughts on doing a hardfork dry-run after the whole scalability
	  thing has cooled off? We could change the address version bytes and
	  payment protocol so the new version is incompatible with old,
	  without touching the consensus layer; this would simulate a hardfork
	  at an economic level (as much as possible without economic losses),
	  and from that hopefully we can learn enough to do a hardfork in the
	  future safely.  [21:10]
<zooko> Luke-Jr: interesting idea. It reminds me of jtimon's proposal at SB/HK
	"Let's have an uncontroversial hard fork!".  [21:14]
<Luke-Jr> zooko: well, hopefully the next hardfork proposal will be
	  uncontroversial anyway - this can help prepare for it ;)
<jtimon> zooko you mean bip99 still in draft waiting for coauthors to just
	 pick up and implement ovious things from the various hardfork
	 wishlists?  [21:18]
* zooko looks at
  https://github.com/bitcoin/bips/blob/master/bip-0099.mediawiki
<jtimon> s/ovious/obvious  [21:19]
<zooko> whoa I need to actually pay attention to this, apparently.
<zooko> Are you telling me that the timewarp attack actually was used against
	altcoins ?
<jtimon> zooko of course it was used against altcoins, that's why almost most
	 altcoins fixed it (probably the fact that most altcoins just forked
	 litecoin was a factor too)
<zooko> jtimon: huh, I thought it wasn't exploitable in practice. So I guess I
	really do need to pay attention! Thanks for mentioning it.
```

**Comment by kanzure** [comment]
http://gnusha.org/bitcoin-wizards/2016-01-29.log

**Comment by zookozcash** [comment]
https://github.com/jtimon/bitcoin/commit/91cceb9ae0061e273865cacc107f49aa6f6879a2

**Comment by ebfull** [comment]
This is the one we want to apply: https://github.com/freicoin/freicoin/commit/beb2fa54745180d755949470466cbffd1cd6ff14

**Comment by daira** [comment]
+1

**Comment by ebfull** [comment]
We get this for free depending on our difficulty adjustment algorithm, right?

**Comment by str4d** [comment]
Yes; this fix only needs to be applied if we simply modify the existing difficulty adjustment algorithm. If we take alternate code then this fix will not apply (but we should still check that the alternate code is not itself vulnerable).

**Comment by defuse** [comment]
Related: #998

**Comment by nathan-at-least** [comment]
Fixed / made-irrelevant in #1004.
